qsc/plot.py

```diff
@@ -0,0 +1,529 @@
+            fieldline_Z[i,j] = Z_2D_spline(phi_mod,theta_fieldline_mod)[0]
+    return fieldline_X, fieldline_Y, fieldline_Z
+
+def create_subplot_mayavi(mlab, R, alphas, x_2D_plot, y_2D_plot, z_2D_plot,
+                          fieldline_X, fieldline_Y, fieldline_Z,
+                          Bmag, degrees_array_x, degrees_array_z, shift_array):
+    '''
+    Plotting routine for a mayavi figure instance that plots
+    both the surface and the field lines together. The number
+    of surfaces to plot is specified by the length of the
+    array degrees_array_x
+
+    Args:
+      mlab: mayavi package
+      R: scipy rotation vector package
+      alphas: array of field line labels alpha
+      x_2D_plot: 2D array for the x components of the surface
+      y_2D_plot: 2D array for the y components of the surface
+      z_2D_plot: 2D array for the z components of the surface
+      fieldline_X: 2D array for the x components of the field line
+      fieldline_Y: 2D array for the x components of the field line
+      fieldline_Z: 2D array for the x components of the field line
+      Bmag: 2D array for the magnetic field modulus on the (theta,phi) meshgrid
+      degrees_array_x: 1D array with the rotation angles in the x direction for each surface
+      degrees_array_z: 1D array with the rotation angles in the z direction for each surface
+      shift_array: 1D array with a shift in the y direction for each surface
+    '''
+    assert len(degrees_array_x) == len(degrees_array_z) == len(shift_array)
+    for i in range(len(degrees_array_x)):
+        # The surfaces and field lines are rotated first in the
+        # z direction and then in the x direction
+        rx= R.from_euler('x', degrees_array_x[i], degrees=True)
+        rz= R.from_euler('z', degrees_array_z[i], degrees=True)
+        # Initialize rotated arrays
+        x_2D_plot_rotated = np.zeros((x_2D_plot.shape[0],x_2D_plot.shape[1]))
+        y_2D_plot_rotated = np.zeros((x_2D_plot.shape[0],x_2D_plot.shape[1]))
+        z_2D_plot_rotated = np.zeros((x_2D_plot.shape[0],x_2D_plot.shape[1]))
+        fieldline_X_rotated = np.zeros((fieldline_X.shape[0],fieldline_X.shape[1]))
+        fieldline_Y_rotated = np.zeros((fieldline_X.shape[0],fieldline_X.shape[1]))
+        fieldline_Z_rotated = np.zeros((fieldline_X.shape[0],fieldline_X.shape[1]))
+        # Rotate surfaces
+        for th in range(x_2D_plot.shape[0]):
+            for ph in range(x_2D_plot.shape[1]):
+                [x_2D_plot_rotated[th,ph], y_2D_plot_rotated[th,ph], z_2D_plot_rotated[th,ph]] = rx.apply(rz.apply(np.array([x_2D_plot[th,ph], y_2D_plot[th,ph], z_2D_plot[th,ph]])))
+        # Rotate field lines
+        for th in range(fieldline_X.shape[0]):
+            for ph in range(fieldline_X.shape[1]):
+                [fieldline_X_rotated[th,ph], fieldline_Y_rotated[th,ph], fieldline_Z_rotated[th,ph]] = rx.apply(rz.apply(np.array([fieldline_X[th,ph], fieldline_Y[th,ph], fieldline_Z[th,ph]])))
```

Comment:
These nested loops over a large number of points might be a rate-limiting step. Could be worth looking at whether this could be vectorized.